sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2Listener.scala

```diff
@@ -0,0 +1,414 @@
+  }
+
+
+  def onSessionCreated(ip: String, sessionId: String, userName: String = "UNKNOWN"): Unit = {
+    postLiveListenerBus(SparkListenerSessionCreated(ip, sessionId,
+      userName, System.currentTimeMillis()))
+  }
+
+  def onSessionClosed(sessionId: String): Unit = {
+    postLiveListenerBus(SparkListenerSessionClosed(sessionId, System.currentTimeMillis()))
+  }
+
+  def onStatementStart(
+    id: String,
+    sessionId: String,
+    statement: String,
+    groupId: String,
+    userName: String = "UNKNOWN"): Unit = {
+    postLiveListenerBus(SparkListenerStatementStart(id, sessionId, statement, groupId,
+      System.currentTimeMillis(), userName))
+  }
+
+  def onStatementParsed(id: String, executionPlan: String): Unit = {
+    postLiveListenerBus(SparkListenerStatementParsed(id, executionPlan))
+  }
+
+  def onStatementCanceled(id: String): Unit = {
+    postLiveListenerBus(SparkListenerStatementCanceled(id, System.currentTimeMillis()))
+  }
+
+  def onStatementError(id: String, errorMsg: String, errorTrace: String): Unit = {
+    postLiveListenerBus(SparkListenerStatementError(id, errorMsg, errorTrace,
+      System.currentTimeMillis()))
+  }
+
+  def onStatementFinish(id: String): Unit = {
+    postLiveListenerBus(SparkListenerStatementFinish(id, System.currentTimeMillis()))
+
+  }
+
+  def onOperationClosed(id: String): Unit = {
+    postLiveListenerBus(SparkListenerOperationClosed(id, System.currentTimeMillis()))
+  }
+
+
+  /** Go through all `LiveEntity`s and use `entityFlushFunc(entity)` to flush them. */
+  private def flush(entityFlushFunc: LiveEntity => Unit): Unit = {
+    sessionList.values.asScala.foreach(entityFlushFunc)
+    executionList.values.asScala.foreach(entityFlushFunc)
+  }
+
+  private def update(entity: LiveEntity, now: Long): Unit = {
+    entity.write(kvstore, now)
+  }
+
+  def updateLiveStore(session: LiveEntity, force: Boolean = false): Unit = {
+    if (live || force == true) {
+      session.write(kvstore, System.nanoTime())
+    }
+  }
+
+  private def getOrCreateSession(
+     sessionId: String,
+     startTime: Long,
+     ip: String,
+     username: String): LiveSessionData = {
+    sessionList.computeIfAbsent(sessionId,
+      (_: String) => new LiveSessionData(sessionId, startTime, ip, username))
+  }
+
+  private def getOrCreateExecution(
+    execId: String, statement: String,
+    sessionId: String, startTimestamp: Long,
+    userName: String): LiveExecutionData = {
+    executionList.computeIfAbsent(execId,
+      (_: String) => new LiveExecutionData(execId, statement, sessionId, startTimestamp, userName))
+  }
+
+  private def cleanupExecutions(count: Long): Unit = {
+    val countToDelete = calculateNumberToRemove(count, retainedStatements)
+    if (countToDelete <= 0L) {
+      return
+    }
+    val view = kvstore.view(classOf[ExecutionInfo]).index("finishTime").first(0L)
+    val toDelete = KVUtils.viewToSeq(view, countToDelete.toInt) { j =>
+      j.finishTimestamp != 0
```

Comment:
deleting executions should not be done based on finishTimestamp, but on
```
  def isExecutionActive(execInfo: ExecutionInfo): Boolean = {
    !(execInfo.state == ExecutionState.FAILED ||
      execInfo.state == ExecutionState.CANCELED ||
      execInfo.state == ExecutionState.CLOSED)
  }
```
(an execution that has finishTimestamp can still be busy fetching data)

Reply:
I think, earlier also it was removing based on `finishedTime`
https://github.com/apache/spark/blob/c98e5eb3396a6db92f2420e743afa9ddff319ca2/sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/HiveThriftServer2.scala#L281-L282

Should I change to `isExecutionActive`?